File: mac/Sources/Padly/Models/Messages.swift
import Foundation

/// All decoded message shapes after Decoder validates the frame.
enum PadlyMessage {
    case hello(Hello)
    case pinAttempt(PinAttempt)
    case ping(ts: UInt64)
    case cursorMove(dx: Float, dy: Float, ts: UInt64)
    case cursorClick(button: MouseButton, down: Bool, ts: UInt64)
    case scroll(dx: Float, dy: Float, momentum: Bool, ts: UInt64)
    case gesture(Gesture)
    case key(KeyEvent)
    case presenter(PresenterAction)
    case bye(reason: String)
}

struct Hello {
    let deviceId: String
    let deviceName: String
    let appVer: String
    let capabilities: [String]
}

struct PinAttempt {
    let hash: Data
    let salt: Data
}

enum MouseButton: UInt8 {
    case left = 0, right = 1, middle = 2
}

struct Gesture {
    enum Phase: UInt8 { case begin = 0, update = 1, end = 2 }
    enum Kind: UInt8 {
        case pinchZoom = 0
        case rotate = 1
        case swipe3Up = 2, swipe3Down = 3, swipe3Left = 4, swipe3Right = 5
        case swipe4Up = 6, swipe4Down = 7, swipe4Left = 8, swipe4Right = 9
        case spread4 = 10, pinch4 = 11
    }
    let kind: Kind
    let phase: Phase
    let data: Data
    let ts: UInt64
}

struct KeyEvent {
    /// HID Usage Page 0x07 code.
    let code: UInt16
    /// Bitmask: cmd=1, opt=2, ctrl=4, shift=8, fn=16.
    let modifiers: UInt8
    let down: Bool
    let ts: UInt64
}

struct PresenterAction {
    enum Kind: UInt8 {
        case next = 0, prev = 1, black = 2, white = 3, pointer = 4, stop = 5
    }
    let kind: Kind
    let pointerXY: CGPoint?
    let ts: UInt64
}

/// Persisted on disk in Keychain.
struct PairedDevice: Identifiable, Codable, Equatable {
    let id: String
    var name: String
    var hmacKey: Data
    var certPin: Data
    var lastSeen: Date
}

/// Modifier bits used by KeyInjector when posting CGEvent.
struct ModifierBits {
    static let cmd: UInt8 = 1
    static let opt: UInt8 = 2
    static let ctrl: UInt8 = 4
    static let shift: UInt8 = 8
    static let fn: UInt8 = 16
}

File: android/app/src/main/java/app/padly/android/proto/Messages.kt
package app.padly.android.proto

import kotlinx.serialization.Serializable

/**
 * Wire-protocol message shapes. Each is a sealed branch of [PadlyMessage].
 * The actual MessagePack encoding lives in [app.padly.android.transport.Encoder];
 * these classes are just the in-process representation.
 */
sealed interface PadlyMessage {
    val ts: ULong

    data class Hello(
        val deviceId: String,
        val deviceName: String,
        val appVer: String,
        val capabilities: List<String>,
        override val ts: ULong = 0u,
    ) : PadlyMessage

    data class PinAttempt(
        val pinHash: ByteArray,
        val salt: ByteArray,
        override val ts: ULong = 0u,
    ) : PadlyMessage {
        override fun equals(other: Any?): Boolean = other is PinAttempt &&
            pinHash.contentEquals(other.pinHash) && salt.contentEquals(other.salt) && ts == other.ts
        override fun hashCode(): Int = pinHash.contentHashCode() * 31 + salt.contentHashCode()
    }

    data class Ping(override val ts: ULong) : PadlyMessage

    data class CursorMove(val dx: Float, val dy: Float, override val ts: ULong) : PadlyMessage
    data class CursorClick(val button: MouseButton, val down: Boolean, override val ts: ULong) : PadlyMessage
    data class Scroll(val dx: Float, val dy: Float, val momentum: Boolean, override val ts: ULong) : PadlyMessage
    data class GestureMsg(val kind: GestureKind, val phase: GesturePhase, val data: ByteArray, override val ts: ULong) : PadlyMessage {
        override fun equals(other: Any?): Boolean = other is GestureMsg && kind == other.kind && phase == other.phase && data.contentEquals(other.data) && ts == other.ts
        override fun hashCode(): Int = kind.hashCode() * 31 + phase.hashCode()
    }
    data class Key(val code: UShort, val modifiers: UByte, val down: Boolean, override val ts: ULong) : PadlyMessage
    data class Presenter(val kind: PresenterKind, val x: Float?, val y: Float?, override val ts: ULong) : PadlyMessage
    data class Bye(val reason: String, override val ts: ULong = 0u) : PadlyMessage
}

enum class MouseButton(val code: UByte) { Left(0u), Right(1u), Middle(2u) }

enum class GesturePhase(val code: UByte) { Begin(0u), Update(1u), End(2u) }

enum class GestureKind(val code: UByte) {
    PinchZoom(0u), Rotate(1u),
    Swipe3Up(2u), Swipe3Down(3u), Swipe3Left(4u), Swipe3Right(5u),
    Swipe4Up(6u), Swipe4Down(7u), Swipe4Left(8u), Swipe4Right(9u),
    Spread4(10u), Pinch4(11u),
}

enum class PresenterKind(val code: UByte) {
    Next(0u), Prev(1u), Black(2u), White(3u), Pointer(4u), Stop(5u),
}

@Serializable
data class PairingPayload(
    val ip: String,
    val port: Int,
    val serverId: String,
    val certFingerprint: String,
    val bootstrapToken: String,
)

object ModifierBits {
    const val CMD: UByte = 1u
    const val OPT: UByte = 2u
    const val CTRL: UByte = 4u
    const val SHIFT: UByte = 8u
    const val FN: UByte = 16u
}
